src/UnoFlipModel.java

```diff
@@ -455,6 +461,20 @@ public ArrayList<TurnSequence> getTurnSeqs() {
         return turnSeqs;
     }
 
+    /**
+     * Return the boolean to see if the self draw one is playable.
+     *
+     * @return return true if valid, otherwise false.
+     */
+    private boolean validSelfDrawOne(){
+        for (int i = 0; i < this.getCurrentPlayer().getHandSize();i++){
+            if (this.getCurrentPlayer().getCard(i).getRank() == this.getCurrentRank() || this.getCurrentPlayer().getCard(i).getColour() == this.getCurrentColour()){
```

Comment:
maybe Card class should have a boolean match(Card c) method that determines if a card can be played or not. This would make this a bit cleaner, and the method might help us in other places too.